src/pudl/transform/ferc1.py

```diff
@@ -1661,108 +2005,261 @@ def _multiplicative_error_correction(tofix, mask, minval, maxval, mults):
-                "expns_boiler": "opex_boiler",
-                "expns_plants": "opex_plants",
-                "expns_misc_steam": "opex_misc_steam",
-                "tot_prdctn_expns": "opex_production_total",
-                "expns_kwh": "opex_per_kwh",
-            }
+def clean_strings_for_table(
+    df: pd.DataFrame, table_name: TABLES_LITERAL
+) -> pd.DataFrame:
+    """Apply ``pudl.helpers.cleanstrings`` to a ferc1 table."""
+    df = pudl.helpers.cleanstrings(
+        df=df,
+        columns=TABLE_CLEAN_STRINGS_COLS.get(table_name).get("columns"),
+        stringmaps=TABLE_CLEAN_STRINGS_COLS.get(table_name).get("stringmaps"),
+        unmapped=TABLE_CLEAN_STRINGS_COLS.get(table_name).get("unmapped", pd.NA),
+    )
+    return df
+
+
+def replace_unknown_w_nulls(df, table_name):
+    """Replace the ``unknown``'s from ``cleanstrings`` with nulls."""
+    df = df.replace(
+        {
+            column: "unknown"
+            for column in TABLE_CLEAN_STRINGS_COLS.get(table_name).get("columns")
+        },
+        pd.NA,
+    )
+    return df
+
+
+def oob_to_nan_for_table(df, table_name):
+    """Apply ``pudl.helpers.oob_to_nan`` to a ferc1 table."""
+    oob_to_nan_inputs = TABLE_OOB_TO_NAN.get(table_name)
+    df = pudl.helpers.oob_to_nan(
+        df=df,
+        cols=oob_to_nan_inputs.get("columns"),
+        lb=oob_to_nan_inputs.get("lower_bound"),
+        ub=oob_to_nan_inputs.get("upper_bound"),
+    )
+    return df
+
+
+##############################################################################
+# DATABASE TABLE SPECIFIC PROCEDURES ##########################################
+##############################################################################
+
+
+def plants_steam_ferc1(ferc1_dbf_raw_dfs, ferc1_xbrl_raw_dfs, ferc1_transformed_dfs):
+    """Transforms FERC Form 1 plant_steam data for loading into PUDL Database.
+
+    Args:
+        ferc1_dbf_raw_dfs: Each entry in this dictionary of DataFrame objects
+            corresponds to a table from the  FERC Form 1 DBC database.
+        ferc1_xbrl_raw_dfs: Each entry in this dictionary of DataFrame objects
+            corresponds to a table from the  FERC Form 1 XBRL database.
+        ferc1_transformed_dfs: dictionary of transformed FERC1 tables (names: dfs).
+            Needed to store output and fuel table is used in assignment of plant
+            IDs.
+    """
+    plants_steam_combo = (
+        pre_concat_clean_and_concat_dbf_xbrl(
+            ferc1_dbf_raw_dfs, ferc1_xbrl_raw_dfs, "plants_steam_ferc1"
         )
-        .pipe(_clean_cols, "f1_steam")
-        .pipe(pudl.helpers.simplify_strings, ["plant_name_ferc1"])
+        .pipe(simplify_strings_for_table, "plants_steam_ferc1")
+        .pipe(clean_strings_for_table, table_name="plants_steam_ferc1")
+        .pipe(oob_to_nan_for_table, table_name="plants_steam_ferc1")
+        .pipe(convert_units_in_table, "plants_steam_ferc1")
+        .pipe(convert_cols_dtypes, data_source="ferc1")
```

Comment:
What I'm seeing here and I think the module more broadly, is some general transformations whose programmatic behavior is controlled by a function with parameters that vary by input table. Those parameters are stored in a bunch of different constants, typically dictionaries keyed by table name.

It seems like we might organize the same information more cleanly with a per-table data class or Pydantic model (say `TableTransformer`) that can contain all of the constants required for transformation, and that has an abstract base class/model that defines the transformation functions as methods, and defines the types of the required transformation parameters. The methods would take a dataframe as input and return a dataframe as output, and they would have access to the required parameters as attributes inside the class.

Each table would have a particular instantiation of this `TableTransformer` class with attributes customized to transform it appropriately. We would store these attributes as dictionaries similar to how we store the Resource and Field definitions now. They could be instantiated by ID. We could either have different versions of the attributes stored inside the same `TableTransformer` for each data source (dbf vs. xbrl) or we could have separate TableTransformer definitions for the different data sources, though if many of the transformations are identical (post-concatenation) for the two data sources, probably that doesn't make sense.

Executing the transformation might look like...

```py
fuel_ferc1_transformer = TableTransformer.from_id("fuel_ferc1")
fuel_ferc1_df = fuel_ferc1_transformer.transform(
    dbf_df=fuel_dbf_ferc1_df,
    xbrl_df=fuel_xbrl_ferc1_df,
)
```

Inside the `TableTransformer.transform()` method the different transformations could be composed into a method chain, which maybe happen in different orders for different tables. If there are bespoke table specific transformations that need to happen, maybe those can be defined in addition to the methods that are available in the base `TableTransformer` class? Or just referred to as external functions which are available in the same module?

Anyway, this is only half-baked, but I think something like this sounds good -- getting all these static metadata-like transformation parameters collected together into a structure that pertains to a given table, with the code that applies them defined in conjunction with that structure.

Reply:
Before we go down this path let me double check it is possible for class methods to be ops.

Reply:
I tried making class methods ops but got dagster errors related to the signature of the op. This was confirmed by a dagster person in slack: 
<img width="632" alt="image" src="https://user-images.githubusercontent.com/17532695/178044159-cf15f73a-0a0c-4b13-8313-822e0aa8e0c6.png">

This means our current design isn't super dagster friendly. The main {table_name}Transform.execute() calls can be wrapped in functions and ops to play nice with Dagster:

```python
def steam_transform_factory(table_name):
    @op
    def steam_transform(raw_dbf: pd.DataFrame,
                        raw_xbrl_instant: pd.DataFrame,
                        raw_xbrl_duration: pd.DataFrame,
                        transformed_fuel: pd.DataFrame,):
        return PlantsSteamFerc1(
            table_name="plants_steam_ferc"
        ).execute(
            raw_dbf=raw_dbf,
            raw_xbrl_instant=raw_xbrl_instant,
            raw_xbrl_duration=raw_xbrl_duration,
            transformed_fuel=transformed_fuel,
        )

# ... inside ferc1.transform()
transformed_steam_df = steam_transform_factory("plants_steam_ferc")(raw_dbf=ferc1_dbf_raw_dfs.get(table),
                                                                    raw_xbrl_instant=ferc1_xbrl_raw_dfs.get(f"{table}_instant", None),
                                                                    raw_xbrl_duration=ferc1_xbrl_raw_dfs.get(f"{table}_duration", None),
                                                                    transformed_fuel=ferc1_transformed_dfs["fuel_ferc1"])
```

This will work but seems a little kludgy. 

I don't think our current plan would enable us to turn generic cleaning functions into ops. 

An alternative would be to retain the `TransformerMeta` class and turn `GenericTransformer` methods into functions. I'll write up some pseudo-code soon.

Reply:
I was going to try and re-organize all of the per-table metadata into being contained by a high-level Pydantic class, with the various recurring attributes of that class which show up in different tables having their own definitions, along the same lines as our nested structures in the broader Package / Resource / Field / etc. structures (which seems to be how Pydantic expects things to work). With a big top-level dictionary keyed by the table name, so we can do something like

```py
ttm_fuel_ferc1 = TableTransformerMeta.from_id("fuel_ferc1")
```

and then that object can be used to organize all of the various parameters that determine transformation behavior.

In an `@op` that knows what table it's being applied to, it would then be possible to access all of that (generally static) information and use it in the transformation, which I guess can just be done by some helper / transform functions instead of a separate class... with each of those functions also having behavior that's largely controlled based on the transformation parameters stored in the Pydantic class for the table they're being applied to.

If it turns out there are some repeating patterns of function application maybe we can compile them into graphs that get re-used?

What's the right way to pass get these transformer helper functions the parameters they need? Should they also just look up based on the Pydantic metadata structure? Or should the individual functions take their more specific parameters? If the latter, then I guess the Pydantic objects would need to be passed in as dictionaries via the config, which could work so long as they never contain data types Pydantic doesn't like.

Or is it better to just assume that all these helper functions are going to get subsumed within the per-table transformation function, and not be something that Dagster is ever aware of?

Reply:
Did this in #1739